### ISSUE TITLE
ReLU(inplace=True) seems something wrong internal

### ISSUE THREAD

**Issue by sonack** [pytorch/pytorch#5687]
With ReLU(inplace=True), my model can not be trained, and its loss goes to hundreds of thousands after a few iterations. However, when I replace with ReLU(inplace=False), all trouble disappear and my loss can be converge gradually. 
Pytorch didn't complain any about my inplace ReLU usage before, everything seemed peace, except the larger and large training loss. I have heard someone said "When pytorch doesnt give you error or warning about your inplace operation usage, it almost be correctly working".  But it seems some wrong internal , how did it happen?

Below is my model code:

```
class ResidualBlock(nn.Module):
    def __init__(self, ch_in, ch_out, shortcut = None):
        super(ResidualBlock, self).__init__()
        self.left = nn.Sequential(
            nn.Conv2d(ch_in, 128, 3, 1, 1),  # ch_in, ch_out, kernel_size, stride, pad
            nn.ReLU(inplace=True),
            nn.Conv2d(128, ch_out, 3, 1, 1),
            nn.ReLU(inplace=True),
        )
        self.right = shortcut
    
    def forward(self, x):
        out = self.left(x)
        residual = x if self.right is None else self.right(x)
        out += residual
        return F.relu(out, inplace=False)

class ContentWeightedCNN(BasicModule):
    def __init__(self):
        super(ContentWeightedCNN, self).__init__()
        self.model_name = 'ContentWeightedCNN'
        self.encoder = self.make_encoder()
        self.decoder = self.make_decoder()
        self.reset_parameters()
    
    def reset_parameters(self):
        self.apply(weights_initialization)
    
    def forward(self, x):
        mgdata = self.encoder(x)
        dec_data = self.decoder(mgdata)
        return dec_data

    def make_encoder(self):
        layers = [
            nn.Conv2d(3, 128, 8, 4, 2),
            nn.ReLU(inplace=True), # 54

            ResidualBlock(128, 128),

            nn.Conv2d(128, 256, 4, 2, 1), # 115
            nn.ReLU(inplace=True),

            ResidualBlock(256, 256),

            nn.Conv2d(256, 256, 3, 1, 1), #192
            nn.ReLU(inplace=True),

            ResidualBlock(256, 256),

            nn.Conv2d(256, 64, 1, 1, 0),    # conv 4  64 is n
            nn.Sigmoid(),                    
        ]
        return nn.Sequential(*layers)

    def make_decoder(self):
        layers = [
            nn.Conv2d(64, 512, 3, 1, 1),
            nn.ReLU(inplace=True),

            ResidualBlock(512, 512),

            nn.Conv2d(512, 512, 3, 1, 1),
            nn.ReLU(inplace=True),

            ResidualBlock(512, 512),

            nn.PixelShuffle(2),

            nn.Conv2d(128, 256, 3, 1, 1),
            nn.ReLU(inplace=True),

            ResidualBlock(256, 256),

            nn.PixelShuffle(4),

            nn.Conv2d(16, 32, 3, 1, 1),
            nn.ReLU(inplace=True),
            
            nn.Conv2d(32, 3, 1, 1, 0)   
        ]
        return nn.Sequential(*layers)
 
            
```

**Comment by fmassa** [comment]
What's you pytorch version? `print(torch.__version__)` from the interpreter should tell you.

**Comment by sonack** [comment]
@fmassa 0.3.0.post4

**Comment by vadimkantorov** [comment]
There are two in-place operations in a row: `nn.ReLU(inplace=True)` and `out += residual` which may be the cause

**Comment by soumith** [comment]
indeed that is the issue as @vadimkantorov pointed out. It's a user-error. nn.ReLU(inplace=True) is not meant to be used with output of any layer that requires it's outputs to compute gradients.

**Comment by sonack** [comment]
@vadimkantorov @soumith  hi, i am still confused about the in-place operations, Could you please explain why `out+=residual `  is a in-place operation to me? i thought in python language view `out` is just ref to another variable which is sum of the previous out and the residual, the correspoding objects still exist in memory, which grad_fn record the backward order, then why it can't be backproped when using in-place in out’s last ReLU? i think the inplace relu only destroy its input , replaced with its output, but  i didn‘t use the input to relu(i.e. the conv output) directly into another branch.so i am really confused, could you please give me some idea about it, thank you all.

**Comment by fmassa** [comment]
@sonack it's an in-place operation because `+=` is the in-place add.
You avoid the in-place by doing
```python
out = out + residual
```
You can verify that with other data structures as well:
```python
a = []
b = a  # b is the same as a
a += [5]
print(b)  # prints [5]

# now
a = []
b = a  # b is the same as a
a = a + [5]
print(b)  # prints []
```

**Comment by fmassa** [comment]
@soumith but there is something fishy here, as this pattern was already present in the [official resnet implementation](https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py#L51-L52), so this might not be the issue here.

**Comment by fmassa** [comment]
Actually, the gradients of `ReLU` are not computed in-place, even if ReLU is in-place ([at least that was the behaviour before merging operations in ATen](https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/thnn/auto.py#L238) I believe). Did this behaviour change over time?

**Comment by apaszke** [comment]
No, we’ve never used inplace functions for gradients. Autograd doesn’t allow that

**Comment by fmassa** [comment]
Ok, so @sonack problem still remains, right? Or was the ResNet model from `torchvision` wrong all the time?

**Comment by apaszke** [comment]
The model definition in `torchvision` should be fine, but what's happening here doesn't look like a user error to me. Not sure why aren't we triggering the version checks.

**Comment by soumith** [comment]
ReLU probably doesn't mark dirty, and hence we aren't triggering version checks.

**Comment by sonack** [comment]
@fmassa You mean pytorch implement the Variable's `__iadd__ `as a in-place add operation while `__add__` as a out-of-place add operation, am I understand right? I haven't seen any information about it in the official tutorial.

**Comment by apaszke** [comment]
@soumith You're right. `threshold_` in `VariableType.cpp` doesn't have a call to `increment_version`. It's a bug in our codegen.

**Comment by fmassa** [comment]
@apaszke great, that might explain why I was having weird convergence problems :) !
But I @sonack is using 0.3.0.post4, which shouldn't be affected by the changes in ATen, is that right?

@sonack yes, that's right. And there is a mention of the drawbacks of in-place operations on Variables in [here])http://pytorch.org/docs/0.3.1/notes/autograd.html#in-place-operations-on-variables).

**Comment by apaszke** [comment]
`post4` is quite recent, so it might be using the new code paths that have this regression. They were used with `Variables` even before the merger.

**Comment by apaszke** [comment]
My bad. `threshold_` basically dispatches to `threshold_forward_`, which bumps the version!

**Comment by weiyangfb** [comment]
@sonack has this issue been resolved yet with 0.4.1 or current master?

**Comment by colesbury** [comment]
Yes, it's fixed. You can't do an inplace-addition after an inplace relu:

```
import torch
relu = torch.nn.ReLU(inplace=True)
x = torch.randn(1024, requires_grad=True)
out = x.clone()
out = relu(out)
out += x
out.sum().backward()
```
```
RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation
```

The code in ResNet is OK because the order is swapped (addition first). Addition doesn't require saving inputs or outputs to compute the gradient. ReLU requires saving the input.
